src/Database/LSMTree/Internal/UniqCounter.hs

```diff
@@ -23,16 +24,17 @@ uniqueToRunNumber :: Unique -> RunNumber
 uniqueToRunNumber = coerce
 
 -- | An atomic counter for producing 'Unique' values.
---
--- TODO: could we use a PrimVar here?
-newtype UniqCounter m = UniqCounter (StrictMVar m Word64)
+newtype UniqCounter m = UniqCounter (PrimVar (PrimState m) Word64)
 
 {-# INLINE newUniqCounter #-}
-newUniqCounter :: MonadMVar m => Word64 -> m (UniqCounter m)
-newUniqCounter x = UniqCounter <$> newMVar x
+newUniqCounter :: PrimMonad m => Word64 -> m (UniqCounter m)
+newUniqCounter x = UniqCounter <$> newPrimVar x
 
 {-# INLINE incrUniqCounter #-}
 -- | Return the current state of the atomic counter, and then increment the
 -- counter.
-incrUniqCounter :: MonadMVar m => UniqCounter m -> m Unique
-incrUniqCounter (UniqCounter uniqVar) = modifyMVar uniqVar (\x -> pure ((x+1), Unique x))
+incrUniqCounter :: PrimMonad m => UniqCounter m -> m Unique
+incrUniqCounter (UniqCounter uniqVar) = do
+  x <- readPrimVar uniqVar
+  modifyPrimVar uniqVar succ
+  pure $ Unique x
```

Comment:
This is not atomic: you'll have to use something like `fetchAddInt`, which requires using `PrimVar s Int`

Moreover, `succ` can fail when go past `maxBound`. Instead, we should preferably just let let the `Int` silently wrap around. Note also how `uniqueToWord64` and `uniqueToRunNumber` are affected by this, because we'll have to cast from `Int` to `Word64`

Reply:
I have updated the code to be atomic, relying on the "under-the-hood" coercions to improve performance and guarantee safety.